wgengine/netstack/netstack.go

```diff
@@ -237,14 +237,24 @@ func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
 	}
 	r.Complete(false)
 	c := gonet.NewTCPConn(&wq, ep)
-	go ns.forwardTCP(c, "100.101.102.103:80")
+	go ns.forwardTCP(c, &wq, "100.101.102.103:80")
 }
 
-func (ns *Impl) forwardTCP(client *gonet.TCPConn, address string) {
+func (ns *Impl) forwardTCP(client *gonet.TCPConn, wq *waiter.Queue, address string) {
 	defer client.Close()
 	ns.logf("[v2] netstack: forwarding to address %s", address)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	waitEntry, notifyCh := waiter.NewChannelEntry(nil)
+	wq.EventRegister(&waitEntry, waiter.EventHUp)
+	defer wq.EventUnregister(&waitEntry)
+	go func() {
+		<-notifyCh
+		cancel()
+	}()
+	// netstack doesn't close the notification channel automatically if there was no
+	// hup signal, so we close it after we're done to not leak the above goroutine.
+	defer close(notifyCh)
```

Comment:
this could cause a double-close panic. You'll probably want to make a separate "done" chan and `defer close(done)` here instead, then change line 252 to a select, waiting for either notifyCh or done.